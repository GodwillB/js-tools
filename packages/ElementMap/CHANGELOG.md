# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.0.5](https://github.com/onaio/js-tools/compare/@onaio/element-map@0.0.2...@onaio/element-map@0.0.5) (2019-05-01)

**Note:** Version bump only for package @onaio/element-map

## [0.0.3](https://github.com/onaio/js-tools/compare/@onaio/element-map@0.0.2...@onaio/element-map@0.0.3) (2019-05-01)

**Note:** Version bump only for package @onaio/element-map

## [0.0.2](https://github.com/onaio/js-tools/compare/@onaio/element-map@0.0.1...@onaio/element-map@0.0.2) (2019-03-21)

**Note:** Version bump only for package @onaio/element-map

## 0.0.1 (2019-03-13)

**Note:** Version bump only for package @onaio/element-map
