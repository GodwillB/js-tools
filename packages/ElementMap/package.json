{
  "name": "element-map",
  "version": "0.0.0",
  "description": "Simple component takes a list of elements and returns them wrapped in the provided HTML tag",
  "main": "ElementMap.js",
  "repository": "https://github.com/onaio/js-tools",
  "author": "Ona Engineering",
  "license": "Apache 2",
  "private": true,
  "scripts": {
    "jest": "jest --coverage --verbose --color"
  },
  "jest": {
    "setupFiles": [
      "../../setupTests"
    ]
  },
  "babel": {
    "presets": [
      "@babel/preset-env",
      "@babel/preset-react"
    ],
    "env": {
      "test": {
        "plugins": [
          "transform-es2015-modules-commonjs"
        ]
      }
    }
  },
  "peerDependencies": {
    "react": "^16.8.1"
  },
  "dependencies": {
    "prop-types": "^15.6.1"
  },
  "devDependencies": {
    "react": "^16.8.1",
    "enzyme": "^3.8.0",
    "enzyme-adapter-react-16": "^1.9.1",
    "enzyme-to-json": "^3.3.5"
  }
}
