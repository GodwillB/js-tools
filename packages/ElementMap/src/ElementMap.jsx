// This component takes a list of elements and returns them wrapped
// in the provided HTML tag
import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ElementMap extends Component {
  render() {
    const { items } = this.props;
    const { className } = this.props;
    const { HTMLTag } = this.props;
    // use React.createElement to create the html tag
    // in a dynamic way
    const listItems = items.map((item, key) =>
      React.createElement(
        HTMLTag,
        /* eslint-disable react/no-array-index-key */
        { key, className },
        /* eslint-enable react/no-array-index-key */
        item
      )
    );

    return listItems;
  }
}

ElementMap.propTypes = {
  items: PropTypes.arrayOf(PropTypes.node).isRequired,
  HTMLTag: PropTypes.string.isRequired,
  className: PropTypes.string
};

ElementMap.defaultProps = {
  className: ''
};

export default ElementMap;
