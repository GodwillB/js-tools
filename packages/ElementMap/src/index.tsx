import React from 'react';

/** Interface for ElementMap props */
export interface ElementMapProps {
  items: any[];
  HTMLTag: string;
  className: string;
}

/** default props for ElementMap */
const defaultElementMapProps: Partial<ElementMapProps> = {
  className: 'element-map'
};

/** A simple component that takes a list of elements and a HTML tag and outputs
 * the same list of elements wrapped in the same HTML tag provided.
 */
class ElementMap extends React.Component<ElementMapProps, {}> {
  public static defaultProps = defaultElementMapProps;

  constructor(props: ElementMapProps) {
    super(props);
  }

  public render() {
    const { HTMLTag, className, items } = this.props;
    /** use React.createElement to create the html tag in a dynamic way */
    const listItems = items.map((item, key) =>
      React.createElement(
        HTMLTag,
        /* eslint-disable react/no-array-index-key */
        { key, className },
        /* eslint-enable react/no-array-index-key */
        item
      )
    );
    return listItems;
  }
}

export default ElementMap;
