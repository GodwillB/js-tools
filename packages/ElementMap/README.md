# ElementMap

A simple `React` component that takes a list of elements and a HTML tag and outputs the same list of elements wrapped in the same HTML tag provided.

## Why is this useful?

`ElementMap` helps to keep code DRY - use it every time that have a list of things that you want to render wrapped inside a HTML tag, which happens a lot of times in any medium/large codebase.

Therefore this (or variations of it):

```js
const items = ['a', 'b', 'c'];
const wrappedItems = items.map(el => <li>el</li>);
return <ol>{wrappedItems}</ol>;
```

Becomes:

```js
import ElementMap from '@onaio/element-map';

const things = ['a', 'b', 'c'];
return (
  <ol>
    <ElementMap items={things} HTMLTag="li" />
  </ol>
);
```

## Examples

```js
import ElementMap from '@onaio/element-map';

const headerItems = ['Name', 'Age'];
let output = <ElementMap items={headerItems} HTMLTag="th" />;
// output should now look like:
// [<th>Name</th>, <th>Age</th>]
```

```js
import ElementMap from '@onaio/element-map';

const headerItems = ['Fruit', 'Price'];
let output = <ElementMap items={headerItems} HTMLTag="th" className="css-class" />;
// output should now look like:
// [<th className="css-class">Fruit</th>, <th className="css-class">Price</th>]
```
