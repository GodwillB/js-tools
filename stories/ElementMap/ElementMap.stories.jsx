// stories of ElementMap
/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { storiesOf } from '@storybook/react';
/* eslint-enable import/no-extraneous-dependencies */
import ElementMap from '../../packages/ElementMap/src/ElementMap';
import notes from '../../packages/ElementMap/README.md';

function renderList() {
  const fruits = ['Apple', 'Banana', 'Grape', 'Orange', 'Plum'];
  return (
    <ul>
      <ElementMap items={fruits} HTMLTag="li" />
    </ul>
  );
}

function renderTable() {
  const fruits = ['Apple', 'Banana', 'Grape', 'Orange', 'Plum'];
  return (
    <table>
      <ElementMap items={fruits} HTMLTag="tr" />
    </table>
  );
}

storiesOf('ElementMap', module)
  .add('list items', renderList, { notes })
  .add('table rows', renderTable, { notes });
