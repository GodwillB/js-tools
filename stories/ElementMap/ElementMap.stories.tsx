// stories of ElementMap
/* eslint-disable import/no-extraneous-dependencies */
import { storiesOf } from '@storybook/react';
import React from 'react';
import notes from '../../packages/ElementMap/README.md';
/* eslint-enable import/no-extraneous-dependencies */
import ElementMap from '../../packages/ElementMap/src/';

function renderList() {
  const fruits = ['Apple', 'Banana', 'Grape', 'Orange', 'Plum'];
  return (
    <ul>
      <ElementMap items={fruits} HTMLTag="li" />
    </ul>
  );
}

function renderTable() {
  const fruits = ['Apple', 'Banana', 'Grape', 'Orange', 'Plum'];
  return (
    <table>
      <tbody>
        <tr>
          <ElementMap items={fruits} HTMLTag="td" />
        </tr>
      </tbody>
    </table>
  );
}

storiesOf('ElementMap', module)
  .add('list items', renderList, { notes })
  .add('table rows', renderTable, { notes });
