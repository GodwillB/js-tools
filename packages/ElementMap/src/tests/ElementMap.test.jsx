// Test Element Map
import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import ElementMap from '../ElementMap';

describe('ElementMap', () => {
  it('renders without crashing', () => {
    const headerItems = ['Name', 'Age'];
    shallow(<ElementMap items={headerItems} HTMLTag="th" />);
  });

  it('renders tags correctly', () => {
    const headerItems = ['Name', 'Age'];
    const headerItems2 = ['Subject', 'Difficulty'];
    const wrapper = mount(<ElementMap items={headerItems} HTMLTag="div" />);
    const wrapper2 = mount(<ElementMap items={headerItems2} HTMLTag="span" className="fancy" />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(toJson(wrapper2)).toMatchSnapshot();
    wrapper.unmount();
  });
});
