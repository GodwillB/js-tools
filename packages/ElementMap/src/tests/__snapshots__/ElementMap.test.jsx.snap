// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElementMap renders tags correctly 1`] = `
<ElementMap
  HTMLTag="div"
  className="element-map"
  items={
    Array [
      "Name",
      "Age",
    ]
  }
>
  <div
    className="element-map"
    key="0"
  >
    Name
  </div>
  <div
    className="element-map"
    key="1"
  >
    Age
  </div>
</ElementMap>
`;

exports[`ElementMap renders tags correctly 2`] = `
<ElementMap
  HTMLTag="span"
  className="fancy"
  items={
    Array [
      "Fruit",
      "Price",
    ]
  }
>
  <span
    className="fancy"
    key="0"
  >
    Fruit
  </span>
  <span
    className="fancy"
    key="1"
  >
    Price
  </span>
</ElementMap>
`;
